src/framing/plaintext.rs

```diff
@@ -47,23 +128,125 @@ impl MLSPlaintext {
+
+        let tbm_payload =
+            MLSPlaintextTBMPayload::new(&tbs_payload, &self.signature, &self.confirmation_tag)?;
+        let membership_tag = MembershipTag::new(ciphersuite, membership_key, tbm_payload)?;
+
+        self.membership_tag = Some(membership_tag);
+        Ok(())
     }
-    pub fn verify(
+
+    /// Verify the signature of an `MLSPlaintext` sent from an external party.
+    /// Returns `Ok(())` if successful or `VerificationError` otherwise.
+    pub fn verify_signature(
         &self,
-        serialized_context_option: Option<Vec<u8>>,
+        serialized_context: &[u8],
         credential: &Credential,
-    ) -> bool {
-        let signature_input = MLSPlaintextTBS::new_from(&self, serialized_context_option);
-        signature_input.verify(credential, &self.signature)
+    ) -> Result<(), VerificationError> {
+        let tbs_payload = MLSPlaintextTBS::new_from(&self, Some(serialized_context));
+        tbs_payload.verify(credential, &self.signature)
+    }
+
+    /// Verify the signature of an `MLSPlaintext` sent from an external party.
+    /// Returns `Ok(())` if successful or `VerificationError` otherwise.
+    pub fn verify_membership_tag(
+        &self,
+        ciphersuite: &Ciphersuite,
+        serialized_context: &[u8],
+        membership_key: &MembershipKey,
+    ) -> Result<(), VerificationError> {
+        let tbs_payload = MLSPlaintextTBS::new_from(&self, Some(serialized_context));
+        let tbm_payload =
+            MLSPlaintextTBMPayload::new(&tbs_payload, &self.signature, &self.confirmation_tag)
+                .map_err(VerificationError::CodecError)?;
+        let expected_membership_tag =
+            &MembershipTag::new(ciphersuite, membership_key, tbm_payload)?;
+
+        // Verify the membership tag
+        if let Some(membership_tag) = &self.membership_tag {
+            if membership_tag != expected_membership_tag {
+                Err(VerificationError::InvalidMembershipTag)
+            } else {
+                Ok(())
+            }
+        } else {
+            Err(VerificationError::MissingMembershipTag)
+        }
     }
+
+    /// Verify the signature and the membership tag of an `MLSPlaintext` sent
+    /// from a group member. Returns `Ok(())` if successful or
+    /// `VerificationError` otherwise.
+    pub fn verify_from_member(
+        &self,
+        ciphersuite: &Ciphersuite,
+        serialized_context: &[u8],
+        credential: &Credential,
+        membership_key: &MembershipKey,
+    ) -> Result<(), VerificationError> {
+        // Verify the signature first
+        let tbs_payload = MLSPlaintextTBS::new_from(&self, Some(serialized_context));
+        tbs_payload.verify(credential, &self.signature)?;
+
+        let tbm_payload =
+            MLSPlaintextTBMPayload::new(&tbs_payload, &self.signature, &self.confirmation_tag)
+                .map_err(VerificationError::CodecError)?;
+        let expected_membership_tag =
```

Comment:
Since we have a `MembershipKey` type, this whole logic should probably be in `membership_key.verify(membership_tag, ...)`.